### PR TITLE
Fixed #30828 -- Added how to remove/insert many-to-many relations in bulk to the database optimization docs.

### PR DESCRIPTION
**Pull request by davidfstr**
Adds documentation explaining how to bulk-add and bulk-remove many-to-many relationships efficiently.

It is written assuming no dedicated methods are added for these operations, such as the operations proposed in PR <https://github.com/django/django/pull/11899> .

---

The code that users are expected to inline is somewhat long, so I still wonder if we might reconsider adding dedicated methods to support the bulk-associate and bulk-disassociate use cases:

* Bulk associate a list of object-pairs
    * 6 lines of boilerplate user code
    * uses the `*`-operator (an advanced Python feature)
    * uses a list comprehension (an advanced Python feature)
* Bulk disassociate a list of object-pairs
    * 6 lines of boilerplate user code
    * uses Q expressions (an advanced Django feature)

### CODE REVIEW

File: docs/topics/db/optimization.txt [hunk 1]
```diff
@@ -250,8 +250,8 @@ Use ``QuerySet.count()``
 Use ``QuerySet.exists()``
 -------------------------
 
-...if you only want to find out if at least one result exists, rather than ``if
-queryset``.
+...if you only want to find out if at least one result exists, rather than 
+``if queryset``.
```
Comment (davidfstr) [inline docs/topics/db/optimization.txt]: Small unrelated change: Reformatted this line slightly so that it didn't break syntax-highlighting in my editor.

File: docs/topics/db/optimization.txt [hunk 2]
```diff
@@ -411,4 +416,104 @@ objects to reduce the number of SQL queries. For example::
...
+Many source objects, many target objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When inserting pairs of objects into :class:`ManyToManyFields
+<django.db.models.ManyToManyField>`, use 
+:meth:`~django.db.models.query.QuerySet.bulk_create()` with multiple
+:attr:`through <django.db.models.ManyToManyField.through>` model instances 
+to reduce the number of SQL queries. For example::
+
+    Pizza_Topping = Pizza.toppings.through
```
Comment (pahaz) [inline docs/topics/db/optimization.txt]: this model naming style may little confuse

Comment (davidfstr) [reply docs/topics/db/optimization.txt]: I'm open to other suggestions.

**Comment by davidfstr** [comment]
@charettes @felixxm In case you didn't see it, this PR is a rewrite of the [older add_relations PR](https://github.com/django/django/pull/11899) that proposes a documentation change instead. Your feedback would again be appreciated.

File: docs/topics/db/optimization.txt [hunk 3]
```diff
@@ -411,4 +416,104 @@ objects to reduce the number of SQL queries. For example::
...
+
+...is preferable to::
+
+    my_pizza.toppings.add(pepperoni)
+    your_pizza.toppings.add(pepperoni, mushroom)
+
+...where ``Pizza`` and ``Topping`` have a many-to-many relationship.
+
+If you have a list of object pairs, use the ``*``-operator when calling
+:meth:`~django.db.models.query.QuerySet.bulk_create()`::
```
Comment (takkaria) [inline docs/topics/db/optimization.txt]: Is this right?  `bulk_create()` takes a list as its first argument so I don't understand why you'd unpack it into positional arguments.

Comment (davidfstr) [reply docs/topics/db/optimization.txt]: Fixed.

File: docs/topics/db/optimization.txt [hunk 4]
```diff
@@ -411,4 +416,105 @@ objects to reduce the number of SQL queries. For example::
     my_band.members.add(me)
     my_band.members.add(my_friend)
 
-...where ``Bands`` and ``Artists`` have a many-to-many relationship.
+...where ``Band`` and ``Artist`` have a many-to-many relationship.
+
+If you have a list of objects, use the ``*``-operator when calling
+:meth:`~django.db.models.fields.related.RelatedManager.add`::
+
+    my_band.members.add(*new_members)
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I don't see much value in describing how to use ``*`` in Python (see [comment](https://github.com/django/django/pull/11597#discussion_r307693994)), please remove these paragraphs.

File: docs/topics/db/optimization.txt [hunk 5]
```diff
@@ -396,9 +396,14 @@ Note that there are a number of :meth:`caveats to this method
 <django.db.models.query.QuerySet.bulk_update>`, so make sure it's appropriate
 for your use case.
 
+.. _insert-many-to-many-in-bulk:
+
 Insert in bulk
 --------------
 
+One source object, many target objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I don't think that we need extra subsections.

File: docs/topics/db/optimization.txt [hunk 6]
```diff
@@ -411,4 +416,105 @@ objects to reduce the number of SQL queries. For example::
...
+    my_band.members.add(*new_members)
+
+Many source objects, many target objects
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When inserting pairs of objects into :class:`ManyToManyFields
+<django.db.models.ManyToManyField>`, use 
+:meth:`~django.db.models.query.QuerySet.bulk_create()` with multiple
+:attr:`through <django.db.models.ManyToManyField.through>` model instances 
+to reduce the number of SQL queries. For example::
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I would say:
```
When inserting different pairs of objects into
:class:`~django.db.models.ManyToManyField` or when the custom
:attr:`~django.db.models.ManyToManyField.through` table is defined, use
:meth:`~django.db.models.query.QuerySet.bulk_create()` method to reduce the
number of SQL queries. For example::

    Membership.objects.bulk_create([
        Membership(group=my_band, person=me, date_joined=date(1962, 8, 16)),
        Membership(group=my_band, person=my_friend, date_joined=date(1960, 8, 1)),
        Membership(group=they_band, person=my_sister, date_joined=date(1971, 8, 1)),
        Membership(group=they_band, person=my_brother, date_joined=date(1972, 1, 1)),
    ])
```

obraz

File: docs/topics/db/optimization.txt [hunk 7]
```diff
@@ -411,4 +416,105 @@ objects to reduce the number of SQL queries. For example::
...
+If you have a list of object pairs, convert them to
+:attr:`through <django.db.models.ManyToManyField.through>`
+model instances first::
+
+    PizzaToppingRelation = Pizza.toppings.through
+    PizzaToppingRelation.objects.bulk_create([
+        PizzaToppingRelation(pizza=pizza, topping=topping)
+        for (pizza, topping) in
+        pizza_topping_tuples
+    ], ignore_conflicts=True)
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I don't think that we need an example how to convert list of objects into list to `through` instances. It seems unnecessary in the advanced topic about optimization.

File: docs/topics/db/optimization.txt [hunk 8]
```diff
@@ -411,4 +416,105 @@ objects to reduce the number of SQL queries. For example::
...
+
+If you have a list of object pairs, you must build the
+:class:`Q <django.db.models.Q>` expression manually::
+
+    from django.db.models import Q
+    PizzaToppingRelation = Pizza.toppings.through
+    pizza_toppings_q = PizzaToppingRelation.objects.none()
+    for (pizza, topping) in pizza_topping_tuples:
+        pizza_toppings_q |= Q(pizza=pizza, topping=topping)
+    PizzaToppingRelation.objects.filter(pizza_toppings_q).delete()
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I would apply the same suggestions to the `Delete in bulk` section.

File: docs/ref/models/relations.txt [hunk 9]
```diff
@@ -46,6 +46,10 @@ Related objects reference
             >>> e = Entry.objects.get(id=234)
             >>> b.entry_set.add(e) # Associates Entry e with Blog b.
 
+        If you have multiple objects to add it is more efficient to use
+        :ref:`a single call to add <insert-many-to-many-in-bulk>`
+        rather than multiple calls.
+
```
Comment (carltongibson) [inline docs/ref/models/relations.txt]: I'm not sure about adding this (or the same for delete below). We already have `*odjs` in the signature, `specified model objects` (plural) in the text above, and uses in the [introductory M2M example](https://docs.djangoproject.com/en/2.2/topics/db/examples/many_to_many/).

It feels a bit out-of-place here. _"more efficient"_ being the exact topic of `optimization.txt`.

File: docs/topics/db/optimization.txt (hunk 6 above)
Comment (carltongibson) [reply docs/topics/db/optimization.txt]: I think for both `Insert` and `Delete` sections a single example is enough. 

I'd be inclined to use `Pizza.toppings.through` in both cases (with `ignore_conflicts=True` for Insert). (Assuming folks with an explicit through model will know to use that...)

File: docs/topics/db/optimization.txt (hunk 7 above)
Comment (davidfstr) [reply docs/topics/db/optimization.txt]: I have removed it, and its paired rather-involved Q() query construction in the bulk-remove section.

However I might highlight: **Now there is now neither documentation for how to explicitly add/remove relations based on a list of object pairs, nor is there a method that provides that functionality.**

The main thrust of the original contribution is no longer present :( , although I do think the docs have been improved in general.

File: docs/ref/models/relations.txt (hunk 9 above)
Comment (davidfstr) [reply docs/ref/models/relations.txt]: I thought having a reference to the optimization section related to these methods would improve their discoverability of the optimization section.

File: docs/topics/db/optimization.txt (hunk 1 above)
Comment (davidfstr) [reply docs/topics/db/optimization.txt]: Done.

File: docs/topics/db/optimization.txt (hunk 5 above)
Comment (davidfstr) [reply docs/topics/db/optimization.txt]: Removed.

File: docs/ref/models/relations.txt (hunk 9 above)
Comment (felixxm) [reply docs/ref/models/relations.txt]: I agree with Carlton, and I'm going to revert this change.

File: docs/topics/db/optimization.txt [hunk 10]
```diff
@@ -411,4 +413,65 @@ objects to reduce the number of SQL queries. For example::
...
+...is preferable to::
+
+    my_pizza.toppings.add(pepperoni)
+    your_pizza.toppings.add(pepperoni, mushroom)
+
+...where ``Pizza`` and ``Topping`` have a many-to-many relationship.
+
+.. _delete-many-to-many-in-bulk:
+
+Delete in bulk
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I renamed this to the `Remove in bulk`.

File: docs/topics/db/optimization.txt [hunk 11]
```diff
@@ -411,4 +413,65 @@ objects to reduce the number of SQL queries. For example::
...
+        PizzaToppingRelation(pizza=your_pizza, topping=pepperoni),
+        PizzaToppingRelation(pizza=your_pizza, topping=mushroom),
+    ], ignore_conflicts=True)
+
+...is preferable to::
+
+    my_pizza.toppings.add(pepperoni)
+    your_pizza.toppings.add(pepperoni, mushroom)
+
+...where ``Pizza`` and ``Topping`` have a many-to-many relationship.
```
Comment (felixxm) [inline docs/topics/db/optimization.txt]: I added a note about `bulk_create()` caveats.

**Comment by felixxm** [comment]
@davidfstr Thanks :+1: I pushed edits.

**Comment by davidfstr** [comment]
Thanks for the reviews folks!
